mgr/cephadm: fix tcmu-runner cephadm_stray_daemon

melissa-kun-li opened the pull request:
Adds tcmu-runner daemon to `managed` to prevent cephadm from raising a cephadm_stray_daemon health warning whenever an image is added to a target

Fixes: https://tracker.ceph.com/issues/51111
Signed-off-by: Melissa Li<melissali@redhat.com>


<!--
Thank you for opening a pull request!  Here are some tips on creating
a well formatted contribution.

Please give your pull request a title like "[component]: [short description]"

This is the format for commit messages:

"""
[component]: [short description]

[A longer multiline description]

Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
Signed-off-by: [Your Name] <[your email]>
"""

The Signed-off-by line is important, and it is your certification that
your contributions satisfy the Developers Certificate or Origin.  For
more detail, see SubmittingPatches.rst.

The component is the short name of a major daemon or subsystem,
something like "mon", "osd", "mds", "rbd, "rgw", etc. For ceph-mgr modules,
give the component as "mgr/<module name>" rather than a path into pybind.

For more examples, simply use "git log" and look at some historical commits.

This was just a quick overview.  More information for contributors is available here:
https://raw.githubusercontent.com/ceph/ceph/master/SubmittingPatches.rst

-->
## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

melissa-kun-li commented:
jenkins test make check

Daniel-Pivonka commented on src/pybind/mgr/cephadm/serve.py:
```diff
@@ -501,7 +501,8 @@ def _check_for_strays(self) -> None:
                                 )
                             )
-
+                    if s.get('type') == 'tcmu-runner':
```
Although this does resolve the stray daemon warning that was happening. This kinda feels weird. Essentially this compares info from list_servers() to the same info from list_servers() so it will never fail the check on line 509. 
I'm not sure where else we would get info about the tcmu-runner containers to make a better comparison? Maybe it would be better to just ignore tcmu-runner containers?

melissa-kun-li replied:
Ah could you explain what you mean by it's comparing info from list_servers() to same info from list_servers()? I thought the issue though was that although `tcmu-runner` is in list_servers(), when I look at `managed` which is from `self.mgr.cache.get_daemon_names()`, I don't see `tcmu-runner`, but it should be considered managed by cephadm. 

```
Nov 06 02:19:37 vm-00 ceph-mgr[18221]: log_channel(cephadm) log [INF] : ls
                                       [{'hostname': 'vm-00', 'services': [{'type': 'mgr', 'id': 'vm-00.tjjcta'}, {'type': 'mon', 'id': 'vm-00'}, {'type': 'osd', 'id': '0'}, {'type': 'tcmu-runner', 'id': 'vm-00:rbd/disk_1'}], 'ceph_version': 'ceph version 17.0.0-8135-gf5b96461 (f5b96461081c2f508d1393202ebcd94e6bd2ea3f) quincy (dev)'}, {'hostname': 'vm-01', 'services': [{'type': 'mgr', 'id': 'vm-01.nzxjid'}, {'type': 'mon', 'id': 'vm-01'}, {'type': 'osd', 'id': '1'}], 'ceph_version': 'ceph version 17.0.0-8135-gf5b96461 (f5b96461081c2f508d1393202ebcd94e6bd2ea3f) quincy (dev)'}, {'hostname': 'vm-02', 'services': [{'type': 'mon', 'id': 'vm-02'}, {'type': 'osd', 'id': '2'}], 'ceph_version': 'ceph version 17.0.0-8135-gf5b96461 (f5b96461081c2f508d1393202ebcd94e6bd2ea3f) quincy (dev)'}]
Nov 06 02:19:37 vm-00 ceph-mgr[18221]: [cephadm INFO cephadm.serve] managed
                                       ['mon.vm-00', 'mgr.vm-00.tjjcta', 'alertmanager.vm-00', 'crash.vm-00', 'grafana.vm-00', 'node-exporter.vm-00', 'osd.0', 'prometheus.vm-00', 'iscsi.iscsi.vm-00.ynpymj', 'crash.vm-01', 'mgr.vm-01.nzxjid', 'mon.vm-01', 'node-exporter.vm-01', 'osd.1', 'iscsi.iscsi.vm-01.svhijq', 'crash.vm-02', 'mon.vm-02', 'node-exporter.vm-02', 'osd.2', 'iscsi.iscsi.vm-02.snrszo']
```

By ignoring tcmu-runner containers do you mean doing something like `if name not in managed and not name.startswith('tcmu-runner'.)`?

adk3798 replied:
@melissa-kun-li Ultimately, this "managed" list is just a temporary list we're using for the purpose of finding stray daemons (before this change it's just equivalent to the daemons in the cache but whatever) so checking if the daemon is in managed or is type tcmu-runner versus just adding all the tcmu-runner daemons to managed are equivalent.  I feel like the point Dan was making is it would be nice to check somehow if the tcmu-runner daemon we find is actually part of one of the iscsi services we managed (or some other check to try to actually verify we actually manage this daemon rather than blanket accepting tcmu-runner daemons as managed). The problem with this is, from the tracker, all we're going to get is tcmu-runner daemons with names like `tcmu-runner.host2.domain.com:iscsi-pool-0001/iscsi-p0001-img-01` and, like Dan,  I don't see what info we could pull from this to try and tell if verify if we "manage" this particular tcmu-runner daemon (other than maybe checking the hostname in the daemon name, but that's already done here via the stray host check). We don't track tcmu-runner daemons in the host cache so there's nothing to compare to. I really can't think of any good check to put in for these tcmu-runner daemons unless we refactor the way we handle the iscsi service so I think blanket accepting them as managed is okay

Tldr, I think this PR is good as soon as the test is added.

melissa-kun-li replied:
Ohh I see what you and Daniel are saying, thanks for explaining Adam! I didn't consider that there may be tcmu-runner daemons outside of the iscsi service we manage. Ah and I think I see what Daniel means by ignoring the tcmu-runner containers. Hmm I'll think if there's a better solution, in the meantime I'll work on the test

sebastian-philipp replied:
ping?

melissa-kun-li commented:
Test on this branch: https://github.com/melissa-kun-li/ceph/commit/30fe25985a1531a18be9d5ece30839b6e0ff7037 just with the symlink change, no PR changes:
https://pulpito.ceph.com/melissali-2021-12-01_15:57:33-rbd:iscsi-master-distro-basic-smithi/
```
2021-12-01T16:28:19.622 INFO:tasks.cram.client.0.smithi026.stdout:/home/ubuntu/cephtest/archive/cram.client.0/gwcli_create.t: failed
2021-12-01T16:28:19.622 INFO:tasks.cram.client.0.smithi026.stdout:--- gwcli_create.t
2021-12-01T16:28:19.623 INFO:tasks.cram.client.0.smithi026.stdout:+++ gwcli_create.t.err
...

2021-12-01T16:28:19.635 INFO:tasks.cram.client.0.smithi026.stdout: Map the LUN
2021-12-01T16:28:19.635 INFO:tasks.cram.client.0.smithi026.stdout: ===========
2021-12-01T16:28:19.635 INFO:tasks.cram.client.0.smithi026.stdout:   $ sudo podman exec $ISCSI_CONTAINER gwcli iscsi-targets/iqn.2003-01.com.redhat.iscsi-gw:ceph-gw/hosts/iqn.1994-05.com.redhat:client disk disk=datapool/block0
2021-12-01T16:28:19.636 INFO:tasks.cram.client.0.smithi026.stdout:+  No such path /iscsi-targets/iqn.2003-01.com.redhat.iscsi-gw:ceph-gw/hosts/iqn.1994-05.com.redhat:client
2021-12-01T16:28:19.636 INFO:tasks.cram.client.0.smithi026.stdout:+  [255]
2021-12-01T16:28:19.636 INFO:tasks.cram.client.0.smithi026.stdout:   $ sudo podman exec $ISCSI_CONTAINER gwcli ls iscsi-targets/ | grep 'o- hosts' | awk -F'[' '{print $2}'
2021-12-01T16:28:19.636 INFO:tasks.cram.client.0.smithi026.stdout:-  Auth: ACL_ENABLED, Hosts: 1]
2021-12-01T16:28:19.636 INFO:tasks.cram.client.0.smithi026.stdout:+  Auth: ACL_ENABLED, Hosts: 0]
2021-12-01T16:28:19.637 INFO:tasks.cram.client.0.smithi026.stdout:   $ sudo podman exec $ISCSI_CONTAINER gwcli ls iscsi-targets/ | grep 'o- iqn.1994-05.com.redhat:client' | awk -F'[' '{print $2}'
2021-12-01T16:28:19.637 INFO:tasks.cram.client.0.smithi026.stdout:-  Auth: None, Disks: 1(300M)]
2021-12-01T16:28:19.637 INFO:tasks.cram.client.0.smithi026.stdout:# Ran 1 tests, 0 skipped, 1 failed.
```

Test on this branch: https://github.com/melissa-kun-li/ceph/commit/21915855678618301f38bb71f03d84f3c3a55e67 (includes PR changes, deleted rbd/iscsi/supported-random-distros, added symlink to single-container-host.yaml): 
https://pulpito.ceph.com/melissali-2021-12-01_06:19:34-rbd:iscsi-master-distro-basic-smithi/6537907/

Same error as the branch without the PR changes.

ideepika commented:
@sebastian-philipp @melissa-kun-li ISCSI gateway creation is failing itself somehow
```
2021-12-01T16:28:19.623 INFO:tasks.cram.client.0.smithi026.stdout:   > IP=`hostname -i | awk '{print $1}'`
2021-12-01T16:28:19.624 INFO:tasks.cram.client.0.smithi026.stdout:   > sudo podman exec $ISCSI_CONTAINER gwcli iscsi-targets/iqn.2003-01.com.redhat.iscsi-gw:ceph-gw/gateways create ip_addresses=$IP gateway_name=$HOST
2021-12-01T16:28:19.624 INFO:tasks.cram.client.0.smithi026.stdout:   $ sudo podman exec $ISCSI_CONTAINER gwcli ls iscsi-targets/ | grep 'o- gateways' | awk -F'[' '{print $2}'
```
@melissa-kun-li can you run this test on ceph master+symlink change, also would be good to have the symlink change in a different PR.

melissa-kun-li commented:
> @melissa-kun-li can you run this test on ceph master+symlink change, also would be good to have the symlink change in a different PR.

Ran with just the symlink change on master using my branch new-rbd-iscsi (https://github.com/melissa-kun-li/ceph/commit/30fe25985a1531a18be9d5ece30839b6e0ff7037) and got the issue in https://github.com/ceph/ceph/pull/43833#issuecomment-983820941

I did:
`teuthology-suite -v --ceph-repo https://github.com/ceph/ceph-ci.git --suite-repo https://github.com/melissa-kun-li/ceph.git -c master -m smithi -s rbd/iscsi -k distro -p 50 --suite-branch new-rbd-iscsi -S 0f9ed11e6727bf4fccef4a0481cd9391f27d3cff --force-priority`

If there's another thing you meant, please let me know

ideepika commented:
> > @melissa-kun-li can you run this test on ceph master+symlink change, also would be good to have the symlink change in a different PR.
> 
> Ran with just the symlink change on master using my branch new-rbd-iscsi ([melissa-kun-li@30fe259](https://github.com/melissa-kun-li/ceph/commit/30fe25985a1531a18be9d5ece30839b6e0ff7037)) and got the issue in [#43833 (comment)](https://github.com/ceph/ceph/pull/43833#issuecomment-983820941)
> 
> I did: `teuthology-suite -v --ceph-repo https://github.com/ceph/ceph-ci.git --suite-repo https://github.com/melissa-kun-li/ceph.git -c master -m smithi -s rbd/iscsi -k distro -p 50 --suite-branch new-rbd-iscsi -S 0f9ed11e6727bf4fccef4a0481cd9391f27d3cff --force-priority`
> 
> If there's another thing you meant, please let me know

Thanks, I will take a look

ideepika commented:
@melissa-kun-li I tried reproducing on master, but could not: http://pulpito.front.sepia.ceph.com/ideepika-2021-12-06_19:34:00-rbd:iscsi-wip-yuval-test1-distro-default-smithi/

melissa-kun-li commented:
@ideepika  thank you for looking into it! I saw your PR: https://github.com/ceph/ceph/pull/44243 . I will try testing again. If we can merge this PR, we can look at the rbd/iscsi tests separately

ideepika commented:
> @ideepika thank you for looking into it! I saw your PR: #44243 . I will try testing again. If we can merge this PR, we can look at the rbd/iscsi tests separately

@melissa-kun-li let me know if you need help with this PR regarding iscsi stuff

sebastian-philipp commented:
https://pulpito.ceph.com/swagner-2022-01-06_10:04:35-orch:cephadm-wip-swagner-testing-2022-01-05-1435-distro-default-smithi/

somehow one of the PRs in this run broke NFS and RGW ingress. I really doubt this PR caused this.

idryomov commented:
It looks like this introduced a bug:

https://tracker.ceph.com/issues/53830

Note the "No such path /iscsi-targets/iqn.2003-01.com.redhat.iscsi-gw:ceph-gw/hosts/iqn.1994-05.com.redhat:client" error that is mentioned in the previous comments.

Melissa's successful run linked in comment https://github.com/ceph/ceph/pull/43833#issuecomment-1005911861 didn't actually test this PR's code because it was scheduled against the master branch.

idryomov commented:
Sorry for the noise, it still fails with this PR taken out.  It's just that nothing seems to be have changed on the RBD side in that date range...
